Replicate osu!(stable)'s hit target position with "Classic" mod

frenzibyte opened the pull request:
- Resolves the final point mentioned in #17422 
- Closes #17422 

I initially thought about resolving this from the legacy piece itself, but after seeing how it all works at an osu!taiko playfield level (hit target occupies a squared area based on the playfield height), I had chose to resolve it via "Classic" mod instead.

| | target |
|-|--------|
| before | <img width="477" alt="CleanShot 2022-04-03 at 02 06 30@2x" src="https://user-images.githubusercontent.com/22781491/161404321-82aefa1e-83fb-4af3-a599-0305057a38a6.png"> |
| after (+ https://github.com/ppy/osu/pull/17621) | <img width="477" alt="CleanShot 2022-04-03 at 02 06 55@2x" src="https://user-images.githubusercontent.com/22781491/161404383-e3ffbe0d-da70-4061-adaa-74a460d74ec9.png"> |
| stable | <img width="477" alt="CleanShot 2022-04-03 at 02 06 58@2x" src="https://user-images.githubusercontent.com/22781491/161404385-ebfe84b3-5784-4929-a84c-814d1ba9923d.png"> |

frenzibyte committed:
Replicate osu!(stable)'s hit target position with "Classic" mod

peppy commented:
There does remaining a question of which one is more correct / better feeling. Did you have any insight into this? I'm not actually sure where the lazer metrics come from, or immediately sure which is better.

frenzibyte commented:
Hard to say, the osu!(lazer) one looks more proper when playing with the current default skin (taking a full square of the playfield for the default hit target piece), that’s why I chose to make this a “Classic” mod thing.

bdach commented:
there's a visual issue here. the reduced distance of the hit target to the input drum causes hit explosions to be clipped on this PR, which is not the case on stable.

| stable | this PR |
| :-: | :-: |
| ![screenshot048](https://user-images.githubusercontent.com/20418176/161429936-0b2dac66-0970-4577-882d-06906f916efe.jpg) | ![osu_2022-04-03_15-12-23](https://user-images.githubusercontent.com/20418176/161429895-e2370447-6e7a-44c9-a26f-bad534cb93a8.jpg) |

peppy commented:
That's likely not going to be an easy one to fix, and may also be the reason it was placed where it was until now...

frenzibyte commented:
Good find, it does look to be quite involved after attempting a couple of failed solutions on it. 

One idea I had in mind is to change the entire proxying logic such that taiko objects are (along with the hit explosion container) by default drawn on top of the input drum, and `ProxyContent` would make it render behind it instead, to achieve the effect of missed objects going beneath the input drum.

Needs more thought as implementing the above idea isn't quite simple with how the taiko playfield is currently structured.

frenzibyte committed:
Merge branch 'master' into taiko-target-classic-position

frenzibyte committed:
Add legacy classic taiko player test scene

frenzibyte committed:
Propagate legacy input drum size to main piece

frenzibyte committed:
Expose input drum `SkinnableDrawable` in `InputDrum` for width consumption

This is probably not a good way to approach this, but I'm unsure about
any other way.

frenzibyte committed:
Reorder taiko playfield elements to fix hit explosion Z-ordering

frenzibyte committed:
Remove redundant code

frenzibyte commented:
Have came up with a solution by masking the bounds of the hit objects container to not potentially render in front of the input drum, and flip depth ordering to the following:
 - Playfield background
 - Input drum (proxy)
 - Mascot
 - Hit explosions and hit target
 - Hit objects and barlines (masked)
 - Drum roll hit container, kiai hit explosions, and judgements
 - Input drum (for consuming input)

So that hit explosions can now render in front of the input drum, while hit objects still render in front of the hit explosions, but get masked away once it exceeds the visible playfield bounds.

![image](https://user-images.githubusercontent.com/22781491/161868949-72e25947-3164-4e16-b51f-dc51c93a8b67.png)

bdach commented on osu.Game.Rulesets.Taiko.Tests/TestSceneLegacyTaikoPlayer.cs:
```diff
@@ -0,0 +1,28 @@
+        [BackgroundDependencyLoader]
+        private void load(SkinManager skins)
+        {
+            skins.CurrentSkinInfo.Value = DefaultLegacySkin.CreateInfo().ToLiveUnmanaged();
```
Can this test scene inherit from [`LegacySkinPlayerTestScene`](https://github.com/ppy/osu/blob/master/osu.Game/Tests/Visual/LegacySkinPlayerTestScene.cs) rather than rolling this on its own?

Would also get rid of the ugly `TestSceneTaikoPlayer` workaround, if I'm interpreting what that's there for correctly.

frenzibyte replied:
Ah, had a feeling a test scene like that existed...

bdach commented on osu.Game.Rulesets.Taiko/Skinning/Legacy/LegacyInputDrum.cs:
```diff
@@ -91,7 +90,7 @@ protected override void Update()
 
             // Relying on RelativeSizeAxes.Both + FillMode.Fit doesn't work due to the precise pixel layout requirements.
             // This is a bit ugly but makes the non-legacy implementations a lot cleaner to implement.
-            content.Scale = new Vector2(DrawHeight / content.Size.Y);
+            Scale = new Vector2(Parent.DrawHeight / Size.Y);
```
This change is taking it from "a bit ugly" to "really ugly". The parent isn't required to really even be there, in general.

Does this need to happen? Can't the input drum just autosize or something, rather than doing all of this moving of the hardcoded size up a level?

bdach commented:
I'll also note that as written this Z-ordering starts applying to the default skin too, which... I don't know if that's good...

![osu_2022-04-06_22-51-17](https://user-images.githubusercontent.com/20418176/162068681-eeb67ca8-728b-41b3-88a6-73da81ade14f.jpg)

If we decide that we want this to be dependent on legacy/non-legacy skin then it may honestly be easier to pull the entire layout of the playfield out to a skinnable or something, because making this conditional in the current state may be painful.

frenzibyte committed:
Use `LegacySkinPlayerTestScene` instead of reimplementing

frenzibyte commented:
Unsure, pulling the entire layout out of the playfield to a skinnable drawable will probably take painful amount of time and effort, and the default skin is bound to be redesigned that I think leaving this as-is might be better, or could tweak the current hit explosion to not extend all over the playfield (cc @peppy).

frenzibyte committed:
Rewrite input drum measurements to autosize on X axis

peppy commented:
> I'll also note that as written this Z-ordering starts applying to the default skin too, which... I don't know if that's good...

I think this is more correct, for sure.

peppy commented on osu.Game.Rulesets.Taiko/UI/TaikoPlayfield.cs:
```diff
@@ -146,6 +163,13 @@ private void load(OsuColour colours)
                 drumRollHitContainer.CreateProxy(),
             };
 
+            // to prioritise receiving key presses on input drum before objects, move input drum to the end of the hierarchy...
+            leftArea.Remove(inputDrum);
+            AddInternal(inputDrum);
+
+            // ...and create a proxy to keep the input drum displayed behind the playfield elements.
+            leftArea.Add(inputDrum.CreateProxy());
```
Can you explain why this is done here and not inline in the above code? Is there some special meaning to adding it to one place then moving it to another?

frenzibyte replied:
Initially had it such that the input drum will be initialized at the the end of the hierarchy, then proxied to the left area in subsequent logic. But simplified that to the following, which interestingly works:
```diff
diff --git a/osu.Game.Rulesets.Taiko/UI/TaikoPlayfield.cs b/osu.Game.Rulesets.Taiko/UI/TaikoPlayfield.cs
index 5891672f29..6d5c7812d5 100644
--- a/osu.Game.Rulesets.Taiko/UI/TaikoPlayfield.cs
+++ b/osu.Game.Rulesets.Taiko/UI/TaikoPlayfield.cs
@@ -65,6 +65,14 @@ private void load(OsuColour colours)
         {
             Container leftArea;

+            inputDrum = new InputDrum(HitObjectContainer)
+            {
+                Anchor = Anchor.CentreLeft,
+                Origin = Anchor.CentreLeft,
+                AutoSizeAxes = Axes.X,
+                RelativeSizeAxes = Axes.Y,
+            };
+
             InternalChildren = new[]
             {
                 new SkinnableDrawable(new TaikoSkinComponent(TaikoSkinComponents.PlayfieldBackgroundRight), _ => new PlayfieldBackgroundRight()),
@@ -74,16 +82,10 @@ private void load(OsuColour colours)
                     RelativeSizeAxes = Axes.Both,
                     FillMode = FillMode.Fit,
                     BorderColour = colours.Gray0,
-                    Children = new Drawable[]
+                    Children = new[]
                     {
                         new SkinnableDrawable(new TaikoSkinComponent(TaikoSkinComponents.PlayfieldBackgroundLeft), _ => new PlayfieldBackgroundLeft()),
-                        inputDrum = new InputDrum(HitObjectContainer)
-                        {
-                            Anchor = Anchor.CentreLeft,
-                            Origin = Anchor.CentreLeft,
-                            AutoSizeAxes = Axes.X,
-                            RelativeSizeAxes = Axes.Y,
-                        },
+                        inputDrum.CreateProxy(),
                     }
                 },
                 mascot = new SkinnableDrawable(new TaikoSkinComponent(TaikoSkinComponents.Mascot), _ => Empty())
@@ -161,15 +163,9 @@ private void load(OsuColour colours)
                     RelativeSizeAxes = Axes.Both,
                 },
                 drumRollHitContainer.CreateProxy(),
+                inputDrum,
             };

-            // to prioritise receiving key presses on input drum before objects, move input drum to the end of the hierarchy...
-            leftArea.Remove(inputDrum);
-            AddInternal(inputDrum);
-
-            // ...and create a proxy to keep the input drum displayed behind the playfield elements.
-            leftArea.Add(inputDrum.CreateProxy());
-
             RegisterPool<Hit, DrawableHit>(50);
             RegisterPool<Hit.StrongNestedHit, DrawableHit.StrongNestedHit>(50);
```

peppy replied:
I'd keep the comment explaining the non-proxy placement alongside it, but seems fine.

peppy commented:
Is this still intentionally blocked? The direction seems fine apart from the pointed out issue.

frenzibyte commented:
Kept this blocked as I still haven't got the chance to test this in depth yet (specifically regarding whether the entire `ProxyContent` logic in `DrawableTaikoHitObject` is necessary anymore, along with whether drumrolls and swells behave as expected with these changes).

frenzibyte committed:
Inline input drum proxying logic

frenzibyte committed:
Merge branch 'master' into taiko-target-classic-position

frenzibyte commented:
On a closer look now that I've found the time to look into this, it seems everything is rendering correctly depth-wise. 

I was wondering whether the `ProxyContent` logic in `DrawableTaikoHitObject` can be removed now that the object always displays in front of the hit target, but didn't realise that I've made the taiko hitobject container masked by default, and that removing the mask requires proxying it, therefore I left it as-is (pretty nice that drawable proxy is able to achieve that). 

This is ready for a second review now, the solution is described in readable words at https://github.com/ppy/osu/pull/17622#issuecomment-1089535770.

peppy committed:
Merge branch 'master' into taiko-target-classic-position

peppy commented:
I think this is okay, but again, I am nearly certain people are going to request the same positional change without classic mod applied. Time will tell I guess.

frenzibyte commented:
> I think this is okay, but again, I am nearly certain people are going to request the same positional change without classic mod applied. Time will tell I guess.

Ah, I've missed that point after looking at the Z-depth issue, but on default skin it still looks somewhat close to the hit target unlike classic skin, so maybe best to leave it as such for now.

peppy commented on osu.Game.Rulesets.Taiko/UI/TaikoPlayfield.cs:
```diff
@@ -86,22 +121,28 @@ private void load(OsuColour colours)
+                                RelativeSizeAxes = Axes.Both,
+                                Children = new Drawable[]
                                 {
-                                    Name = "Hit objects",
-                                    RelativeSizeAxes = Axes.Both,
-                                    Children = new Drawable[]
-                                    {
-                                        HitObjectContainer,
-                                        drumRollHitContainer = new DrumRollHitContainer()
-                                    }
-                                },
+                                    barLinePlayfield = new BarLinePlayfield(),
+                                    HitObjectContainer,
+                                }
+                            }
```
This broke barlines displaying outside of the playfield area.
